### DISCUSSION TITLE
RFC: v1.19 full-A multi-factor modeling and momentum research

### DISCUSSION THREAD

**Discussion by ZacharyHu0** [ZacharyHu0/QuantMaster#398] [General]
# QuantMaster v1.19：全 A 股多因子模型研究与动量增强

> 状态：规划中（canonical planning document）  
> Milestone：[1.19.0](https://github.com/ZacharyHu0/QuantMaster/milestone/16)  
> Project：[QuantMaster v1.19 Quant Research](https://github.com/users/ZacharyHu0/projects/5)  
> 当前阶段：只建立规划记录，不开始实现、不创建分支或 PR。

## 背景与目标

QuantMaster 当前的量化因子主要以单因子验证和固定多因子合成为主。v1.19 的目标是把 Quant Lab 深化为可复现、可审计的全 A 股多因子 Model Research，并把中期动量提升为第一等研究模板。

完成后，一次正式 Study 必须满足以下结果之一：

1. 产出带完整数据、特征、训练、OOF、封存集和生命周期证据的候选；
2. 明确说明依赖、数据、预算、模型失败或统计门槛为何阻止候选产生。

不得静默降级、静默删除模型，也不得把旧证据解释成 v3 合同。

## 模块边界 / Seam

Model Research 对应用层只公开三个领域接口：

```python
plan(request, frozen_snapshot) -> StudyPlan
run(study_id, plan, control) -> StudyEvidence
predict(candidate_ref, frozen_inputs) -> PredictionBundle
```

- `StudyPlan` 使用 v3、内容寻址且不可变；冻结输入快照、因子版本、模型集合、研究协议、目标、预算和执行配置。
- `StudyEvidence` 保存试验、共同 OOF、FDR、封存集、执行成本、失败、检查点与产物谱系。
- `PredictionBundle` 输出预期超额/绝对收益、上涨概率、成本后为正概率、分位数、不确定性、OOD 与降级状态。
- UI、CLI 和任务运行器只依赖上述接口；enqueue/list/人工审批/Champion 仍属于应用层。
- 七个后端使用封闭的内部映射；不建立插件注册表或通用 artifact DAG。

## 全 A 股数据与标签

- 正式 universe 为每个交易日按本地证券目录上市/退市生命周期重建的全 A 股集合。
- warm-up 不足的标的排除；研究结束日超过目录可证明日期时明确失败。
- 不用当前成分、当前行业或当前 ST 状态重写历史；无法证明历史 ST 时不推断。
- 停牌、真实涨跌停与 T+1 只按当时可知证据进入执行模拟。
- stockdb 与本地缓存优先；远端提供者仅补正式必需且本地缺失或陈旧的字段。
- 训练超额标签定义为：个股未来收益减去当日完整合格全 A 股未来收益中位数。必须先用完整集合计算标签，再进行第一阶段抽样。
- 沪深300、中证500、中证1000只作为可勾选的报告基准，默认三条全选；不改变训练标签或 Champion 排名。

## 因子与动量模板

自定义 Study 可钉住 3–20 个已验证因子版本。所有版本 ID、定义和谱系必须冻结；拒绝重复/循环谱系以及 `learned`、`latent` 嵌套输入。

### medium_term_momentum_core_v1

- 20/60/120/252 日原始动量；
- 60–5、120–20、252–20 跳近期动量；
- 期限结构、加速度、风险调整、趋势斜率/质量、上涨日一致性；
- 52 周高点距离与回撤；
- 成交量、成交额、换手率确认和市场状态控制；
- 原始动量与涨跌停事件拆解并存，包括可交易动量、近窗事件次数和价格距离。

### medium_term_momentum_enriched_v1

在 core 之上增加：

- point-in-time 行业相对动量与行业/风格残差动量；
- 本地可审计的异常成交额、换手、成交量与涨停注意力代理；
- 本地新闻事件强度与情绪。

每个开发折和封存折的行业与资讯摄取水位共同 PIT 覆盖率必须达到 90%。不足时阻塞 enriched Study，不填充、不静默回退 core。core 与 enriched 使用独立 Study 和排行榜。

## 七模型两阶段锦标赛

正式 Study 固定比较：

- 线性：Ridge、Elastic Net；
- 树：HistGradientBoosting、RandomForest；
- 深度：Transformer、TCN、GRU。

### 第一阶段：筛选面板调参

- 每个目标日按交易所和截至当日的 20 日成交额三档分层；
- 使用冻结 seed/date/symbol 哈希抽取最多 1,024 只；不足时使用全部合格标的并记录；
- `linear_trials/tree_trials/deep_trials` 可在计划创建时配置并冻结，默认 4/6/8；
- 每个模型使用独立、版本化搜索空间，不把模型类型混入一个分类搜索变量。

### 第二阶段：完整全 A 股共同评估

- 七个模型分别携带第一阶段优胜配置；
- 在完整合格全 A 股上重新训练并生成共同开发期 OOF；
- 第二阶段不继续调参；
- 开发排行榜冻结后只让冠军打开一次封存集；
- 冠军封存失败时该 Study 无候选，不重新调参、不换亚军。

任何必跑模型无法获得成功配置或完成共同 OOF，正式 Study 都不得产生候选。

## 协议、目标与预算

- 正式输出 horizon：5/10/20/30/60d；
- `primary_horizon`：20d 或 60d，默认 20d；
- `top_n`：默认 20，随 StudyPlan 冻结；
- `rebalance_days`：5/10/20，默认 5，随 StudyPlan 冻结；
- 研究目标：`net_ir` 或 `rank_ic`，默认成本后 `net_ir`；
- 默认研究协议：756 日训练、244 日测试/步长、3 个开发折、60 日 purge、末尾 244 日封存；
- 全部已计分 trial/model/horizon 属于同一 Study 的 BH-FDR 假设族；
- 保留覆盖率、折间符号稳定、q≤0.10、IC/ICIR 和 edge/cost 门槛；
- 单次预算默认 10h、上限 24h；启动前估算依赖、设备、缓存、磁盘和预计时间；
- 中断恢复不得重置累计预算；预算耗尽未完成则无候选。提高预算必须创建新的 StudyPlan。

`quantmaster[ml]` 继续承载 scikit-learn、PyTorch 与 Optuna。桌面发布包不内置这些依赖；缺少所选后端时预检明确阻塞并提供安装/设备检查动作。

## 生命周期与 Champion

- 20d 与 60d 各最多一个 Champion；
- rebalance、Top-N、模型家族、动量模板和 Study 目标都作为 Challenger 配置竞争，不再拆分 Champion 槽；
- 封存门通过后自动成为 Shadow Challenger；
- Shadow 至少需要 `max(20, 2 × primary_horizon)` 个成熟信号日，且成本后超额为正、回撤与覆盖门通过；
- Paper 至少运行 20 个交易日，成本后收益非负且无持续订单/数据异常；
- Shadow → Paper → Champion 均需人工操作、操作者和理由；
- 跨研究目标统一按生产指标替换：净 IR 至少提高 0.10，或 IR 差异不超过 0.10 且回撤降低至少 10%、净收益不下降；
- 没有匹配 v3 Champion 时使用规则基线；旧 learned Champion 不自动迁移。

v2 Study 只保留历史摘要，不恢复、不预测、不回测、不晋升。旧动作必须返回明确的 schema unsupported 错误，不提供转换器或兼容解码器。

## Model Research 工作台

用一个 Model Research 工作台替换现有“模型实验”和“滚动优化”，继续使用当前原生 HTML/CSS/JS 技术栈：

- 模板入口：Momentum Core、Momentum Enriched、Custom；
- 流程：数据/依赖预检 → 冻结计划 → 运行阶段 → 七模型排行榜 → 封存证据 → 晋升轨道；
- 状态：loading、empty、error、blocked、paused、no-candidate、legacy-read-only；
- 支持 390px 布局、键盘导航、明确焦点、触摸目标与 reduced-motion。

## 非目标

v1.19 本轮不引入：

- 投资组合线性优化器；
- XGBoost、LightGBM、CatBoost；
- 外部社交媒体热度；
- 券商实盘连接；
- 旧 Study/模型转换器；
- 新插件系统；
- ML 依赖内置桌面包；
- 版本文件、CHANGELOG、tag 或 GitHub Release。

## 执行任务图

依赖关系：

`全 A 数据 → v3 契约 → {七模型锦标赛, 动量模板} → {Model Research UI, 生命周期集成}`

- Epic：[Issue #399](https://github.com/ZacharyHu0/QuantMaster/issues/399)
- 全 A 数据：[Issue #400](https://github.com/ZacharyHu0/QuantMaster/issues/400)
- v3 契约：[Issue #401](https://github.com/ZacharyHu0/QuantMaster/issues/401)
- 七模型锦标赛：[Issue #402](https://github.com/ZacharyHu0/QuantMaster/issues/402)
- 动量模板：[Issue #403](https://github.com/ZacharyHu0/QuantMaster/issues/403)
- Model Research UI：[Issue #404](https://github.com/ZacharyHu0/QuantMaster/issues/404)
- 生命周期集成：[Issue #405](https://github.com/ZacharyHu0/QuantMaster/issues/405)

所有任务初始为 Backlog；只有 RFC 决策完整且依赖满足后才进入 Ready。

### 2026-08-18 · 初始规划冻结

- 线性规划解释为线性预测模型，不做组合优化；
- 正式七模型全部进入完整全 A 股第二阶段；
- 第二阶段冻结第一阶段配置，不继续追参；
- 主预测期可选 20/60d，默认 20d；
- 调仓期可选 5/10/20 日，默认 5 日；
- Champion 只按主预测期分两个槽，Top-N 与调仓周期参与竞争；
- Discussion 作为当前唯一规范文档，仓库 Markdown 延后到另行授权的文档 PR。

后续每项变更先以评论记录“问题、选项、决定、理由、受影响 Issue”，再折叠进本正文。

## 研究依据

- [Returns to Buying Winners and Selling Losers: Implications for Stock Market Efficiency](https://onlinelibrary.wiley.com/doi/10.1111/j.1540-6261.1993.tb04702.x)
- [Time Series Momentum](https://pages.stern.nyu.edu/~lpederse/papers/TimeSeriesMomentum.pdf)
- [The 52-Week High and Momentum Investing](https://onlinelibrary.wiley.com/doi/abs/10.1111/j.1540-6261.2004.00695.x)
- [Price limits and momentum in the Chinese stock market](https://www.sciencedirect.com/science/article/abs/pii/S0264999322001560)
